File: lua/lexer.lua
local token_definitions = {
	{ "NUMBER",     "^(%-?%d+%.?%d*)" },
	{ "HASH",       "^(#)" },
	{ "LABEL",      "^([%a_][%w_]*):" },
	{ "SYMBOL",     "^([%a_][%w_]*)" },
	{ "LPAREN",     "^(%()" },
	{ "RPAREN",     "^(%))" },
	{ "COMMA",      "^(,)" },
	{ "WHITESPACE", "^(%s+)" },
}

local tokenize = function(input)
	local tokens = {}
	local pos = 1

	while pos <= #input do
		local matched = false
		local substring = input:sub(pos)

		for _, def in ipairs(token_definitions) do
			local type, pattern = def[1], def[2]
			local start, finish, capture = substring:find(pattern)

			if start then
				if type ~= "WHITESPACE" then
					table.insert(tokens,
						{ type = type, value = capture or substring:sub(start, finish) })
				end

				pos = pos + finish
				matched = true
				break
			end
		end

		if not matched then
			error("Unexpected character at position " .. pos .. ": " .. input:sub(pos, pos))
		end
	end

	return tokens
end

return tokenize

File: go/lexer.go
package subleq

import (
	"fmt"
	"regexp"
)

type Token struct {
	Type  string `json:"type"`
	Value string `json:"value"`
}

type tokenDef struct {
	Type    string
	Pattern *regexp.Regexp
	Group   int
}

var tokenDefinitions = []tokenDef{
	{"NUMBER", regexp.MustCompile(`^-?\d+\.?\d*`), 0},
	{"HASH", regexp.MustCompile(`^#`), 0},
	{"LABEL", regexp.MustCompile(`^([a-zA-Z_][a-zA-Z0-9_]*):`), 1},
	{"SYMBOL", regexp.MustCompile(`^[a-zA-Z_][a-zA-Z0-9_]*`), 0},
	{"LPAREN", regexp.MustCompile(`^\(`), 0},
	{"RPAREN", regexp.MustCompile(`^\)`), 0},
	{"COMMA", regexp.MustCompile(`^,`), 0},
	{"WHITESPACE", regexp.MustCompile(`^\s+`), 0},
}

func Tokenize(input string) ([]Token, error) {
	var tokens []Token
	pos := 0

	for pos < len(input) {
		matched := false
		substring := input[pos:]

		for _, def := range tokenDefinitions {
			loc := def.Pattern.FindStringSubmatchIndex(substring)
			if loc != nil && loc[0] == 0 {
				if def.Type != "WHITESPACE" {
					value := ""
					if def.Group > 0 && len(loc) > (def.Group*2+1) {
						value = substring[loc[def.Group*2]:loc[def.Group*2+1]]
					} else {
						value = substring[loc[0]:loc[1]]
					}
					tokens = append(tokens, Token{Type: def.Type, Value: value})
				}
				pos += loc[1]
				matched = true
				break
			}
		}

		if !matched {
			return nil, fmt.Errorf("unexpected character at position %d: %c", pos, input[pos])
		}
	}

	return tokens, nil
}
